### PR TITLE
First draft on Streaming Developer Guide

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Add Stream Introduction
   - Spring Cloud Stream & SCDF interactions
   - Binding properties
   - Binder properties

 - Standalone Stream development & testing
  - On RabbitMQ, Apache Kafka

 - Create OOTB stream
 - Creation of custom stream application

 - Continuous Delivery of streaming applications

 - Add stream features

Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3096
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3098
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3172
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3173

### CODE REVIEW

File: .idea/workspace.xml [hunk 1]
```diff
@@ -0,0 +1,882 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cppwfs) [inline .idea/workspace.xml]: need to update .gitignore to remove .idea files.   (found this out hard way as well :-) ).

**Comment by cppwfs** [comment]
Can't pull down the PR because of the .idea/ directory as well as shown here:

```
error: The following untracked working tree files would be overwritten by merge:
	.idea/workspace.xml
Please move or remove them before you merge.
```

**Comment by ilayaperumalg** [comment]
@cppwfs Removed the `.idea` directory. Let me know if this works now.

**Comment by cppwfs** [comment]
@ilayaperumalg WOOHOO it works!

File: content/documentation/3-stream-developer-guides/3-programming-models/3-kafka-streams.md [hunk 2]
```diff
@@ -6,25 +6,125 @@ description: 'Kafka Streams Processing Developer Guide'
 
 # Kafka Streams Processing
```
Comment (sabbyanandan) [inline content/documentation/3-stream-developer-guides/3-programming-models/3-kafka-streams.md]: @sobychacko is looking to add programming model guides for SI, Flux, KStreams, and Functions [see: https://github.com/spring-cloud/spring-cloud-dataflow/issues/3095]. The goal was to hopefully showcase the business logic in the 4 models. Perhaps we can remove this section from the commit until we decide what business logic we want to showcase.

File: content/documentation/3-stream-developer-guides/4-continuous-delivery/1-cd-basics.md [hunk 3]
```diff
@@ -6,34 +6,267 @@ description: 'Continuous Delivery with Skipper'
...
+
+## Stream Rollback
+
+If you want to roll back the event stream to a specific version, you can use the command `stream rollback http-events-transformer --releaseVersion <release-version>`.
+
+After rolling back to the initial version of the event stream (where the `transform` application just did uppercase conversion):
+
+```
+dataflow:>stream rollback http-events-transformer --releaseVersion 1
+Rollback request has been sent for the stream 'http-events-transformer'
```
Comment (sabbyanandan) [inline content/documentation/3-stream-developer-guides/4-continuous-delivery/1-cd-basics.md]: Maybe also show the `stream history` output after rollback.

File: content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+We will start from initializr and create two Spring Cloud Stream applications.
+Note for CF we need a manifest, for k8s we need a service/deployment yaml.
+
+## Development
+
+### Initialzr
+
+Go to [Spring Cloud Stream Initializr](https://start-scs.cfapps.io/) and
+
+- Generate a `Maven Project` With Spring Boot `2.1.4` (Choose the version of your choice)
```
Comment (cppwfs) [inline content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Try to avoid specifying a specific release i.e. 2.1.4. For example `Spring Boot current release`

File: content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+- Select Group `com.example`
+- Select Artifact `demo-source`
+- Search for `RabbitMQ binder starter` dependency
+- Click `Generate Project`
+
```
Comment (cppwfs) [inline content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Include a link so that the user can click a preconfigured zip file .  Shown here in the batch PR: https://github.com/spring-io/dataflow.spring.io/pull/6/files#diff-dc94a800788a54f1beaeb53b17fd8e31R43

File: content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+### Biz Logic
+
+Let's create a domain model `UsageDetail` that contains `userId`, `data` and `duration`.
+
+```java
+public class UsageDetail {
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: In this case we can link it to file in the github repo and save some space.   ie. https://github.com/spring-io/dataflow.spring.io/pull/6/files#diff-7143d3f49a21ce3ab3a5fa6764271b9cR78

File: content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+		this.dataCost = dataCost;
+	}
+
+	public String toString() {
+		return "{\"userId\": \""+this.getUserId() + "\", \"callCost\": \""+ this.getCallCost()+"\", \"dataCost\": \"" + this.getDataCost()+ "\" }";
+	}
+}
+```
+
+The `Source` application produces usage detail for each user including call duration, data usage.
```
Comment (cppwfs) [inline content/documentation/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Include a link to the file in github

File: content/documentation/3-stream-developer-guides/2-streams/3-data-flow-stream.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+Deploy the stream:
+
+```
+dataflow:> stream deploy usage-cost-logger
+```
+
+Once the stream is deployed on `Local` development environment, you can look the runtime applications via Dashboard's runtime page or using the SCDF Shell command `runtime apps`.
+The runtime applications show information about where each application is running in the local environment and their log files locations.
+
```
Comment (cppwfs) [inline content/documentation/3-stream-developer-guides/2-streams/3-data-flow-stream.md]: Not sure if we need to show the log info here.  
But if we do we need show them a screenshot of where to obtain the log file location from runtime apps and then an example bash shell of them catting or tailing the log.

File: content/documentation/3-stream-developer-guides/4-continuous-delivery/1-cd-basics.md [hunk 9]
```diff
@@ -6,34 +6,267 @@ description: 'Continuous Delivery with Skipper'
...
-## Update the stream
+### Registering out of the box streaming applications using the Shell
 
-### Shell
+```
+dataflow:> app import --uri  http://bit.ly/Einstein-SR2-stream-applications-kafka-maven
+```
 
-https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#spring-cloud-dataflow-stream-lifecycle-update
+The above SCDF shell command registers the `Maven` artifacts of the out of the box streaming applications that can work on `RabbitMQ` messaging middleware.
```
Comment (cppwfs) [inline content/documentation/3-stream-developer-guides/4-continuous-delivery/1-cd-basics.md]: This says rabbitmq, but the examples show kafka.

File: content/documentation/5-feature-guides/1-streams/0-index.md [hunk 10]
```diff
@@ -6,3 +6,123 @@ summary: true
 ---
 
 Discuss the set of features that will be covered.
```
Comment (cppwfs) [inline content/documentation/5-feature-guides/1-streams/0-index.md]: This line is causing the generated page to not render properly.  Needs to be moved below the meta-tags.

File: content/documentation/3-stream-developer-guides/1-getting-started/1-simple-stream.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+The stream DSL `http --server.port=9000 | log` can represent the streaming pipeline that has `http` source application ingesting http events into `log` sink application.
+
+The property `server.port` is the `http` application properties set at the stream creation.
+
+## UI
+
+Screen shots here
+
+### Shell
```
[the remark this answers is not in the document]
Comment (ilayaperumalg) [reply content/documentation/3-stream-developer-guides/1-getting-started/1-simple-stream.md]: I think we decided to include both to reach out to both the UI/Shell audience.

**Comment by cppwfs** [comment]
@ilayaperumalg we need to remove the .idea files from the PR.  The PR is failing to be pulled due to the following error:
```
error: The following untracked working tree files would be overwritten by merge:
	.idea/codeStyles/codeStyleConfig.xml
	.idea/dataflow.spring.io.iml
	.idea/misc.xml
	.idea/modules.xml
	.idea/vcs.xml
	.idea/workspace.xml
Please move or remove them before you merge.
Aborting
```

**Comment by ilayaperumalg** [comment]
@cppwfs Sorry, it was my mistake when I pushed the changes without noticing the `.idea` directory (though I added this into .gitignore). Pushed the changes to remove the .idea directory now.

**Comment by cppwfs** [comment]
No worries at all!  Thanks for fixing!

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+1. In the Dependencies text box, type `Rabbitmq` to select the RabbitMQ binder dependency.
+1. In the Dependencies text box, type `Cloud Stream` to select the Spring Cloud Stream dependency.
+1. Click the Generate Project button.
+1. Unzip the usage-cost-processor.zip file and import the project into your favorite IDE.
+
+##### Initialzr Shortcut
+
+Another option instead of using the UI to initialize your project you can do the following:
+
+1. Click the [here](https://start.spring.io/starter.zip?fakeusernameremembered=&fakepasswordremembered=&type=maven-project&language=java&bootVersion=2.1.4.RELEASE&baseDir=usage-cost-processor&groupId=io.spring.dataflow.sample&artifactId=usage-cost-processor&name=usage-cost-processor&description=Sample+project+for+Spring+Cloud+Stream+Processor&packageName=com.example.demosource&packaging=jar&javaVersion=1.8&inputSearch=&style=amqp&style=cloud-stream) to download the preconfigured usage-cost-processor.zip.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: The package name needs to be updated to the groupid.

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+2. Unzip the usage-cost-processor.zip file and import the project into your favorite IDE
+
+#### Biz Logic
+
+1.  In your favorite IDE create the `io.spring.dataflow.sample.domain` package
+1.  Create a `UsageDetail` class in the `io.spring.dataflow.sample.domain` using your favorite IDE that looks like the contents in [UsageDetail.java](https://github.com/spring-cloud/spring-cloud-dataflow-samples/blob/master/dataflow-website/stream-developer-guides/streams/usage-detail-sender/src/main/java/com/example/demo/domain/UsageDetail.java).
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: UsageDetail link returns a 404

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+2. Unzip the usage-cost-processor.zip file and import the project into your favorite IDE
+
+#### Biz Logic
+
+1.  In your favorite IDE create the `io.spring.dataflow.sample.domain` package
+1.  Create a `UsageDetail` class in the `io.spring.dataflow.sample.domain` using your favorite IDE that looks like the contents in [UsageDetail.java](https://github.com/spring-cloud/spring-cloud-dataflow-samples/blob/master/dataflow-website/stream-developer-guides/streams/usage-detail-sender/src/main/java/com/example/demo/domain/UsageDetail.java).
+    This `UsageDetail` model contains `userId`, `data` and `duration`.
+1.  Create a `UsageCostDetail` class in the `io.spring.dataflow.sample.domain` using using your favorite IDE that looks like the contents in [UsageCostDetail.java](https://github.com/spring-cloud/spring-cloud-dataflow-samples/blob/master/dataflow-website/stream-developer-guides/streams/usage-cost-processor/src/main/java/com/example/demo/domain/UsageCostDetail.java).
+    This `UsageCostDetail` model contains `userId`, `callCost` and `dataCost`.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: UsageCostDetail  link returns 404

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+1.  Create the `Processor` application that receives the `UsageDetail` from the previously created `source`, computes the call/data cost and returning the `UsageCostDetail`.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Its unclear that we want to create a class. May want to reword this to something like:
"In the io.spring.dataflow package create a UsageCostProcessor class using your favorite IDE that looks like the contents below." It's unclear where to create the class as well.

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    UsageCostDetail usageCostDetail = new UsageCostDetail();
+    usageCostDetail.setUserId(usageDetail.getUserId());
+    usageCostDetail.setCallCost(usageDetail.getDuration() * this.ratePerSecond);
+    usageCostDetail.setDataCost(usageDetail.getData() * this.ratePerMB);
+    return usageCostDetail;
+    }
+    }
+    ```
+
+#### Testing
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: We should have a section on creating the Unit test for all examples

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+```
+spring.cloud.stream.bindings.input.destination=test-usage-detail
+spring.cloud.stream.bindings.output.destination=test-usage-cost
+server.port=9091
+```
+
+You can run the standalone `UsageCostProcessor` processor application as,
+
+```
+java -jar usage-cost-processor-0.0.1-SNAPSHOT.jar --spring.cloud.stream.bindings.input.destination=test-usage-detail --spring.cloud.stream.bindings.output.destination=test-usage-cost --server.port=9091
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Following these instructions.  The app starts and then terminates with no exception.  I may have missed something though.

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+Another option instead of using the UI to initialize your project you can do the following:
+
+1. Click the [here](https://start.spring.io/starter.zip?fakeusernameremembered=&fakepasswordremembered=&type=maven-project&language=java&bootVersion=2.1.4.RELEASE&baseDir=usage-detail-sender&groupId=io.spring.dataflow.sample&artifactId=usage-detail-sender&name=usage-detail-sender&description=Sample+project+for+Spring+Cloud+Stream+Source&packageName=io.spring.dataflow.sample&packaging=jar&javaVersion=1.8&inputSearch=&style=amqp&style=cloud-stream) to download the preconfigured usage-detail-sender.zip.
+
+2. Unzip the usage-detail-sender.zip file and import the project into your favorite IDE
+
+#### Biz Logic
+
+1.  In your favorite IDE create the `io.spring.dataflow.sample.domain` package
+1.  Create a `UsageDetail` class in the `io.spring.dataflow.sample.domain` using your favorite IDE that looks like the contents in [UsageDetail.java](https://github.com/spring-cloud/spring-cloud-dataflow-samples/blob/master/dataflow-website/stream-developer-guides/streams/usage-detail-sender/src/main/java/com/example/demo/domain/UsageDetail.java).
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: UsageDetail link shows a 404

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+#### Biz Logic
+
+1.  In your favorite IDE create the `io.spring.dataflow.sample.domain` package.
+1.  Create a `UsageCostDetail` class in the `io.spring.dataflow.sample.domain` using using your favorite IDE that looks like the contents in [UsageCostDetail.java](https://github.com/spring-cloud/spring-cloud-dataflow-samples/blob/master/dataflow-website/stream-developer-guides/streams/usage-cost-processor/src/main/java/com/example/demo/domain/UsageCostDetail.java).
+    This `UsageCostDetail` model contains `userId`, `callCost` and `dataCost`.
+1.  Create the `Sink` application that receives the `UsageCostDetail` from the previously created `processor` and logs it.
+
+```java
+@EnableBinding(Sink.class)
+public class UsageCostLogger {
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Its unclear that we want to create a class. May want to reword this to something like:
"In the io.spring.dataflow package create a UsageCostLogger class using your favorite IDE that looks like the contents below." It's unclear where to create the class as well.

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+
+```
+spring.cloud.stream.bindings.input.destination=test-usage-cost
+server.port=9092
+```
+
+You can run the standalone `UsageCostLogger` sink application as,
+
+```
+java -jar usage-cost-logger-0.0.1-SNAPSHOT.jar --spring.cloud.stream.bindings.input.destination=test-usage-cost --server.port=9092
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: app started and then stopped, using the instructions above.

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+1. In the Dependencies text box, type `Rabbitmq` to select the RabbitMQ binder dependency.
+1. In the Dependencies text box, type `Cloud Stream` to select the Spring Cloud Stream dependency.
+1. Click the Generate Project button.
+1. Unzip the usage-detail-sender.zip file and import the project into your favorite IDE.
+
+##### Initialzr Shortcut
+
+Another option instead of using the UI to initialize your project you can do the following:
+
+1. Click the [here](https://start.spring.io/starter.zip?fakeusernameremembered=&fakepasswordremembered=&type=maven-project&language=java&bootVersion=2.1.4.RELEASE&baseDir=usage-detail-sender&groupId=io.spring.dataflow.sample&artifactId=usage-detail-sender&name=usage-detail-sender&description=Sample+project+for+Spring+Cloud+Stream+Source&packageName=io.spring.dataflow.sample&packaging=jar&javaVersion=1.8&inputSearch=&style=amqp&style=cloud-stream) to download the preconfigured usage-detail-sender.zip.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: Link needs to be updated using kafka instead of AMQ.

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+    		usageDetail.setDuration(new Random().nextInt(300));
+    		usageDetail.setData(new Random().nextInt(700));
+    		this.source.output().send(MessageBuilder.withPayload(usageDetail).build());
+    	}
+    }
+    ```
+
+#### Testing
+
+We can individually test these custom applications before creating a pipeline using Spring Cloud Data Flow.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: Needs to be updated for Kafka

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+java -jar usage-detail-sender-0.0.1-SNAPSHOT.jar --spring.cloud.stream.bindings.output.destination=test-usage-detail --server.port=9090
+```
+
+### Sample Processor
+
+#### Initialzr
+
+1. Visit the [Spring Initialzr site](https://start.spring.io/).
+1. Select the latest release of spring boot.
+1. Create a new Maven project with a Group name of `com.example` and an Artifact name of `usage-cost-processor`.
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: com.example needs to be io.spring.dataflow.sample

File: content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md [hunk 24]
```diff
@@ -0,0 +1,222 @@
+---
+path: 'stream-developer-guides/streams/standalone-stream-kafka/'
```
Comment (cppwfs) [inline content/documentation/pages/3-stream-developer-guides/2-streams/2-standalone-stream-kafka.md]: Update instructions to be Kafka instead of RabbitMQ.

File: content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md (hunk 20 above)
Comment (markpollack) [reply content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: the pom has kafka in it as well....not rabbit.  There needs to be another directory level, one for kafka, one for rabbit,  `stream-developer-guides/streams/standalone-stream-kafka` and `stream-developer-guides/streams/standalone-stream-rabbitmq`

Comment (markpollack) [reply content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: Did you add the code from the biz logic step?

Comment (markpollack) [reply content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: The link for UsagecostDetail is also wrong.

Comment (cppwfs) [reply content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: yes.   
```
package io.spring.dataflow.sample;

import io.spring.dataflow.sample.domain.UsageCostDetail;
import io.spring.dataflow.sample.domain.UsageDetail;

import org.springframework.cloud.stream.annotation.EnableBinding;
import org.springframework.cloud.stream.annotation.StreamListener;
import org.springframework.cloud.stream.messaging.Processor;
import org.springframework.messaging.handler.annotation.SendTo;

@EnableBinding(Processor.class)
public class UsageCostProcessor {

	private double ratePerSecond = 0.1;

	private double ratePerMB = 0.05;

	@StreamListener(Processor.INPUT)
	@SendTo(Processor.OUTPUT)
	public UsageCostDetail processUsageCost(UsageDetail usageDetail) {
		UsageCostDetail usageCostDetail = new UsageCostDetail();
		usageCostDetail.setUserId(usageDetail.getUserId());
		usageCostDetail.setCallCost(usageDetail.getDuration() * this.ratePerSecond);
		usageCostDetail.setDataCost(usageDetail.getData() * this.ratePerMB);
		return usageCostDetail;
	}
}
```

Comment (ilayaperumalg) [reply content/documentation/pages/3-stream-developer-guides/2-streams/1-standalone-stream-rabbitmq.md]: I refactored the code by moving the package in Github but missed to update the link. will fix.

**Comment by markpollack** [comment]
merged as 7cfa4f25fa0a73eb19243a667425a026fa368375
